src/Compilers/CSharp/Test/Emit/CodeGen/CodeGenAsyncDisposableTests.cs

```diff
@@ -0,0 +1,1192 @@
+            {
+                return;
+            }
+        };
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces);
+            comp.VerifyDiagnostics(
+                // (8,26): error CS4034: The 'await' operator can only be used within an async lambda expression. Consider marking this lambda expression with the 'async' modifier.
+                //             using await (var y = new C())
+                Diagnostic(ErrorCode.ERR_BadAwaitWithoutAsyncLambda, "var y = new C()").WithArguments("lambda expression").WithLocation(8, 26)
+                );
+        }
+
+        [Fact]
+        public void TestInAsyncAnonymousMethod()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    C()
+    {
+        System.Console.Write(""C "");
+    }
+    public static void Main()
+    {
+        System.Action x = async () =>
+        {
+            using await (var y = new C())
+            {
+                System.Console.Write(""body "");
+                return;
+            }
+        };
+        x();
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        System.Console.Write(""DisposeAsync"");
+        return System.Threading.Tasks.Task.CompletedTask;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, options: TestOptions.DebugExe);
+            comp.VerifyDiagnostics();
+            CompileAndVerify(comp, expectedOutput: "C body DisposeAsync");
+        }
+
+        [Fact]
+        public void TestInUnsafeRegion()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    async System.Threading.Tasks.Task<int> M<T>()
+    {
+        unsafe
+        {
+            using await (var x = new C())
+            {
+                return 1;
+            }
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces, options: TestOptions.UnsafeDebugDll);
+            comp.VerifyDiagnostics(
+                // (8,26): error CS4004: Cannot await in an unsafe context
+                //             using await (var x = new C())
+                Diagnostic(ErrorCode.ERR_AwaitInUnsafeContext, "var x = new C()").WithLocation(8, 26)
+                );
+        }
+
+        [Fact]
+        public void TestInLock()
+        {
+            string source = @"
+class C : System.IAsyncDisposable
+{
+    async System.Threading.Tasks.Task<int> M<T>()
+    {
+        lock(this)
+        {
+            using await (var x = new C())
+            {
+                return 1;
+            }
+        }
+    }
+    public System.Threading.Tasks.Task DisposeAsync()
+    {
+        throw null;
+    }
+}
+";
+            var comp = CreateCompilationWithMscorlib46(source + s_interfaces);
```

Comment:
>CreateCompilationWithMscorlib46 [](start = 23, length = 31)

I see this is the only runtime we're testing here. Is this the only runtime where that type is available?

Reply:
Yes, if I used `CreateStandardCompilation`, most tests would get something like:
```C#
                // (12,26): error CS1061: 'Task' does not contain a definition for 'GetAwaiter' and no extension method 'GetAwaiter' accepting a first argument of type 'Task' could be found (are you missing a using directive or an assembly reference?)
                //             using await (var y = new C())
                Diagnostic(ErrorCode.ERR_NoSuchMemberOrExtension, "var y = new C()").WithArguments("System.Threading.Tasks.Task", "GetAwaiter").WithLocation(12, 26),
                // (23,44): error CS0117: 'Task' does not contain a definition for 'CompletedTask'
                //         return System.Threading.Tasks.Task.CompletedTask;
                Diagnostic(ErrorCode.ERR_NoSuchMember, "CompletedTask").WithArguments("System.Threading.Tasks.Task", "CompletedTask").WithLocation(23, 44)
```